app/assets/javascripts/discourse/app/lib/uppy-chunked-upload.js

```diff
@@ -0,0 +1,351 @@
+import {
+  AbortController,
+  createAbortError,
+} from "@uppy/utils/lib/AbortController";
+
+const MB = 1024 * 1024;
+
+const defaultOptions = {
+  limit: 5,
+  retryDelays: [0, 1000, 3000, 5000],
+  getChunkSize() {
+    return 5 * MB;
+  },
+  onStart() {},
+  onProgress() {},
+  onChunkComplete() {},
+  onSuccess() {},
+  onError(err) {
+    throw err;
+  },
+};
+
+/**
+ * Used mainly as a replacement for Resumable.js, using code cribbed from
+ * uppy's S3 Multipart class, which we mainly use the chunking algorithm
+ * and retry/abort functions of. The _buildFormData function is the one
+ * which shapes the data into the same parameters as Resumable.js used.
+ *
+ * See the UppyChunkedUploader class for the uppy uploader plugin which
+ * uses UppyChunkedUpload.
+ */
+export default class UppyChunkedUpload {
+  constructor(file, options) {
+    this.options = {
+      ...defaultOptions,
+      ...options,
+    };
+    this.file = file;
+
+    if (!this.options.getChunkSize) {
+      this.options.getChunkSize = defaultOptions.getChunkSize;
+      this.chunkSize = this.options.getChunkSize(this.file);
+    }
+
+    this.abortController = new AbortController();
+    this.chunksInProgress = 0;
+    this.chunks = null;
+    this.chunkState = null;
+
+    this._initChunks();
+  }
+
+  _aborted() {
+    return this.abortController.signal.aborted;
+  }
+
+  _initChunks() {
+    const chunks = [];
+
+    if (this.file.size === 0) {
+      chunks.push(this.file.data);
+    } else {
+      for (let i = 0; i < this.file.data.size; i += this.chunkSize) {
+        const end = Math.min(this.file.data.size, i + this.chunkSize);
+        chunks.push(this.file.data.slice(i, end));
+      }
+    }
+
+    this.chunks = chunks;
+    this.chunkState = chunks.map(() => ({
+      uploaded: 0,
+      busy: false,
+      done: false,
+    }));
+  }
+
+  _createUpload() {
+    if (this._aborted()) {
+      throw createAbortError();
+    }
+    this.options.onStart();
+    this._uploadChunks();
+  }
+
+  _uploadChunks() {
+    if (this.chunkState.every((state) => state.done)) {
+      this._completeUpload();
+      return;
+    }
+
+    // For a 100MB file, with the default min chunk size of 5MB and a limit of 10:
+    //
+    // Total 20 chunks
+    // ---------
+    // Need 1 is 10
+    // Need 2 is 5
+    // Need 3 is 5
+    const need = this.options.limit - this.chunksInProgress;
+    const completeChunks = this.chunkState.filter((state) => state.done).length;
+    const remainingChunks = this.chunks.length - completeChunks;
+    let minNeeded = Math.ceil(this.options.limit / 2);
+    if (minNeeded > remainingChunks) {
+      minNeeded = remainingChunks;
+    }
+    if (need < minNeeded) {
+      return;
+    }
+
+    const candidates = [];
+    for (let i = 0; i < this.chunkState.length; i++) {
+      const state = this.chunkState[i];
+      // eslint-disable-next-line no-continue
+      if (state.done || state.busy) {
+        continue;
```

Comment:
Since you already have an `if` statement here, why not avoid continue as eslint recommends by negating the condition and putting your logic in the block/scope?